### PR TITLE
[MAINTENANCE]  Refactor ParameterBuilder classes to extend parent class where possible; also, minor cleanup

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-464/GREAT-498/GREAT-655
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/usage_statistics/usage_statistics.py [hunk 1]
```diff
@@ -336,7 +336,7 @@ def usage_statistics_wrapped_method(*args, **kwargs):
                 time_end: int = int(round(time.time() * 1000))
                 delta_t: int = time_end - time_begin
 
-                handler = get_usage_statistics_handler(args)
+                handler = get_usage_statistics_handler(list(args))
```
Comment (cdkini) [inline great_expectations/core/usage_statistics/usage_statistics.py:339]: why is this necessary? don't *args come in the form of a list already?

Comment (alexsherstinsky) [reply great_expectations/core/usage_statistics/usage_statistics.py:339]: @cdkini PyCharm was issuing wrong type error.

File: great_expectations/rule_based_profiler/config/base.py [hunk 2]
```diff
@@ -427,39 +427,44 @@ def resolve_config_using_acceptable_arguments(
...
+            rule.name: rule for rule in effective_rules
+        }
+        runtime_rules: Dict[str, dict] = {
+            name: RuleBasedProfilerConfig._substitute_variables_in_config(
+                rule=rule,
+                variables_container=effective_variables,
             )
-            runtime_rules[name] = rule_with_substituted_vars
+            for name, rule in effective_rules_dict.items()
+        }
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/config/base.py:453]: I'd request we keep this a for loop and remove the comprehension. It greatly helps with debugging and is easier to read.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/config/base.py:453]: @cdkini it just felt to keep in the same pattern I had in other places -- short, clear, and elegant -- seemed to come natural :)

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 3]
```diff
@@ -23,6 +24,9 @@
 )
 from great_expectations.validator.metric_configuration import MetricConfiguration
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/helpers/util.py:28]: ❤️

File: great_expectations/rule_based_profiler/config/base.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/config/base.py:453]: This is an acceptable usage of a comprehension but I'd request that we use only use it in cases that it makes sense.

There are other instances in our codebase that are wildly complicated and unwieldy. It may be elegant but readability should be a higher priority in my opinion.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/config/base.py:453]: @cdkini I agree that we should think twice before going ahead with using complex (not simple) double-comprehensions and even single ones.  The question becomes what is complex.  To me the barrier becomes when the right hand side is some complicated method with many arguments, but I admit it is a bit of a fuzzy boundary.  I know that a triple-comprehension is too much for me, but most double ones can be rationalized as being on the elegant side.  Not sure.  I do find, however, some single comprehensions, if "institutionalized" in a given module, become idiomatic, which I then want to copy for consistency purposes.  This one just "begged" to be folded into a quick and tight comprehension (especially since the right hand side has a great unit test).  Happy to discuss.  Thanks.

File: great_expectations/rule_based_profiler/rule_based_profiler.py [hunk 4]
```diff
@@ -1221,3 +1199,23 @@ def list_profilers(
...
+
+    :param builder_config: candidate builder override configuration
+    :raises: ProfilerConfigurationError
+    """
+    if not all(
+        [
+            isinstance(builder_config, dict),
+            "class_name" in builder_config,
+            "module_name" in builder_config,
+        ]
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Nitpick - can we make this a generator by using parens instead?

**Review by cdkini — APPROVED** [review]
LGTM! I'll finally concede on your point about "reconciliation as dict" 😆 

Don't love the extra work but I see the value you're providing with it.

File: great_expectations/rule_based_profiler/config/base.py (hunk 2 above)
Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/config/base.py:453]: @alexsherstinsky I do see your point, but also find myself doing a bit of mental gymnastics here. If we are going to keep the code, could we add an inline comment?

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 5]
```diff
@@ -355,6 +359,26 @@ def build_simple_domains_from_column_names(
...
+            "variables"
+        ]
+    except (TypeError, KeyError) as e:
+        logger.warning("Could not convert existing variables to dict: %s", e)
+
+    if variables_dict is None:
+        variables_dict = {}
+
+    return variables_dict
+
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/helpers/util.py:380]: ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
Love it :) Thank you @alexsherstinsky

File: great_expectations/rule_based_profiler/rule_based_profiler.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: @cdkini Committed (among a few other conversions from list to generators) -- thanks!

File: great_expectations/rule_based_profiler/config/base.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/config/base.py:453]: @Shinnnyshinshin Sorry, just saw this -- for the benefit of the release, let us add the comment later.

File: great_expectations/rule_based_profiler/rule_based_profiler.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler.py]: Big 🙇 to @cdkini with huge thanks for thoughtful and balanced reviews.

**Review by NathanFarmer — APPROVED** [review]
LGTM as well!
